034-kafka-topic-replication-factor.md

```diff
@@ -0,0 +1,76 @@
+
+Apache Kafka allows users to change the RF for any topic partition through the native scripts during the runtime. However, it's not a lightweight process, because users usually need to decide which brokers to host new replicas and which replicas need to be dropped, or even conduct a cluster rebalance after the change of the RF.  
+
+The goal of this proposal is to offer users a simple approach to change the topic RF through the KafkaTopic Custom Resources and Cruise Control, in the cloud native, strimzi style. 
+
+## Current situation
+
+Currently, users have two options to change the topic RF:
+* Leverage the Native Kafka script to change the RF. Users need to generate the partition replica assignment and balance the kafka clusters manually if needed. 
+* Delete the topic and recreate the topic with the desired RF.
+
+Clearly none of the above approaches is user-friendly.
+
+When users attempt to change the RF in the KafkaTopic Resources, the Status of the KafkaTopic will become `NotReady` with `ReplicationFactorChangeException`. The following error message is shared with users.
+
+`Changing 'spec.replicas' is not supported. This KafkaTopic's 'spec.replicas' should be reverted to 4 and then the replication should be changed directly in Kafka.`
+
+## Motivation
+
+With this new feature, it will become much easier to achieve the desired tradeoff between the cost and data reliability in the kafka clusters. 
+
+For instance, users can lower the RFs for selective less critical topics to reduce the infrastructure cost (Compute, Storage and Network). Also, for some critical kafka topics, users can increase the data reliability through the bump of the RFs ONLY for these specific topics. 
+
+More importantly, users only need to specify the new RF number in the `KafkaTopic` resource and approve the change, in order to change the RF for kafka topics. The heavy lifting work, such as new replica placement, extra replica deletion and possible cluster rebalance, is completely taken care by the strimzi. 
+
+## Proposal
+
+### API
+The Kafka Topic RF can be changed by users in the following steps:
+* Update the `spec.replicas` properties in the target `KafkaTopic` Resource in the kubernetes cluster
+* Watch the `Status` field of the `KafkaTopic Resource`, for the status update of this RF change.
+
+### Design
+We propose to change the topic RFs with the support of `Cruise Control` and `Strimzi Topic Operator`. 
+
+#### Cruise Control
+Cruise Control Service will be leveraged to change the RF for topics in the kafka clusters, for the following reasons: 
+* Strimzi supports the Cruise Control Service, and it's ready for use with simple configurations. 
+* Cruise Control has native support for the RF changes. It automates the whole process, including selecting of new replicas and extra replicas, triggering the replica creation in the kafka brokers and monitoring the cluster during the process.
```

Comment:
what does this mean `new replicas and extra replicas` ? what do you mean by "extra replicas"?

Reply:
For the `extra replicas`, I mean the replicas that we need to remove. For the `new replica`, I meant the replicas we need to create.

For instance, if a user wants to change RF from 4 to 3. The current replica is [101, 102,103,104]. The cruise control may pick the replica [104] and delete it from the Kafka Brokers. So the 104 is the `extra replica`. 

I think `obsolete` is a better term than the `extra`. I will change it in this proposal.